ChildMindInstitute/mindlogger-backend-refactor
# M2-7011: extend the list of supported item types

### 📝 Description
🔗 [Jira Ticket M2-7013](https://mindlogger.atlassian.net/browse/M2-7013)
Allow create new Item types:
[Date](https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/88735752/AUS-7.2.2.2+-+Item+Types+dropdown#:~:text=5-,Date,-(Mobile%20Only)):
* greater than;
* less than;
* is equal to;
* is not equal to;
* between;
* outside of;

[Number Selection](https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/88735752/AUS-7.2.2.2+-+Item+Types+dropdown#:~:text=6-,Number%20Selection,-Age%20Selector):
* greater than;
* less than;
* is equal to;
* is not equal to;
* between;
* outside of;

[Time](https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/88735752/AUS-7.2.2.2+-+Item+Types+dropdown#:~:text=7-,Time,-(Mobile%20Only)):
* greater than;
* less than;
* is equal to;
* is not equal to;
* between;
* outside of;

[Time Range](https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/88735752/AUS-7.2.2.2+-+Item+Types+dropdown#:~:text=8-,Time%20Range,-(Mobile%20Only)):
* greater than;
* less than;
* is equal to;
* is not equal to;
* between;
* outside of;

[Single Selection per Row](https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/88735752/AUS-7.2.2.2+-+Item+Types+dropdown#:~:text=10-,Single%20Selection%20per%20Row,-(Mobile%20Only)):
* is equal to;
* is not equal to;

[Multiple Selection per Row](https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/88735752/AUS-7.2.2.2+-+Item+Types+dropdown#:~:text=11-,Multiple%20Selection%20per%20Row,-(Mobile%20Only)):
* includes;
* doesn’t include;

[Slider Rows](https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/88735752/AUS-7.2.2.2+-+Item+Types+dropdown#:~:text=12-,Slider%20Rows,-(Mobile%20Only)):
* greater than;
* less than;
* is equal to;
* is not equal to;
* between;
* outside of;

More context about the ticket:
https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/596410369/Item+Flow+-+Extend+the+list+of+supported+Item+Types


### 🪤 Peer Testing
Assuming your Backend and Admin project is running and configured.

How to test via insomnia or postman:
You should be able to request the new item types, such as: GREATER_THAN_DATE, ...
I will attach a json for import a collection of insomnia, so you can use as a Base ( remember to update with your valid informations like applets, id's, the collection is just a sample of body structure of the request.)
[requests.json](https://github.com/user-attachments/files/17104960/requests.json)
**STEPS TO TEST ON INSOMNIA OR POSTMAN**
1 - first open admin side,
2 - open or create an applet
3 - create an item that was mentioned in description as a new item type ( dont forget to pull the item to the top of hierarchy, that is, this way you can select another fields on condition) 
4 - go to the tab item_flow and create a add condition
5 - inspect the page and go to network tab.
6 - create the new condition while you are watching the network, this way you will be able to catch the request
7 - the request loo
![Screenshot 2024-09-23 at 3 08 41 PM](https://github.com/user-attachments/assets/87aa490d-38b9-4ec2-9d86-1dd40c459ed2)
ks like:
8 - go to payload, right click and copy as object and now you can put this object on your insomnia or postman and remember to set up the right applet id for request on postman or insomnia, there is many ways to get the id, the faster way is getting on URL.
![Screenshot 2024-09-23 at 3 11 04 PM](https://github.com/user-attachments/assets/6d5c5818-dd66-49bb-9302-ee65f9d15899)

**STEPS TO TEST ON WEB ADMIN**
1 - create a new applet or open an existing one
2 - create a new item type that was described on description ( dont forget to pull the item to the top of hierarchy, that is, this way you can select another fields on condition) 
3 - go to item flow tab and create a new condition with this item.
4 - validate if the Item and condition will be successfully created, example:
![Screenshot 2024-09-23 at 3 19 12 PM](https://github.com/user-attachments/assets/f483952b-f5e7-4216-8cf3-a1f46cd106a8)

### ✏️ Notes
this Branch was created from this implementation:
https://github.com/ChildMindInstitute/mindlogger-backend-refactor/pull/1419

for more details about the task you can access:
https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/596410369/Item+Flow+-+Extend+the+list+of+supported+Item+Types
